[BUG] Microphone disconnection leaves audio stream stuck, causing silent failures and preventing new recordings

### **Bug Description**

When a USB microphone temporarily disconnects or drops hardware connection during an active session (or right at computer startup), Handy does not handle the stream drop gracefully.

**What happens:**

1. The physical mic disconnects mid-recording or drops connection.


2. The recording fails silently and outputs `sample count: 0`.


3. Upon stopping, the app fails to transcribe (`stop() failed: sending on a closed channel`) because no samples were retrieved.


4. **App lock-up state:** For subsequent recording attempts, the app remains stuck believing the old stream is still active (`Microphone stream already active` -> `Failed to start recording: Recorder not available`).


5. The user cannot record again until the internal stream timeout finishes and resets the audio manager (approx. 1–2 minutes later).



**Expected Behavior:**
If an active microphone disconnects, Handy should detect the broken stream, reset the stream flag gracefully, prompt/notify the user, and re-initialize the audio device (or switch back to system default) without locking up subsequent record hotkey actions.

---

### **System Information**

* **App Version:** `0.9.3`

* **Operating System:** `Windows 11`

* **CPU:** `AMD Ryzen 7`

* **GPU:** `NVIDIA GeForce GTX 1660 Ti`


---

### **Logs**

```text
1. Stream Disconnection & Recorder Lock-up State:
[2026-07-20][06:10:09][handy_app_lib::audio_toolkit::audio::recorder][ERROR] Stream error: The requested device is no longer available. For example, it has been unplugged.
[2026-07-20][06:10:12][handy_app_lib::actions][DEBUG] TranscribeAction::start called for binding: transcribe
[2026-07-20][06:10:12][handy_app_lib::managers::audio][DEBUG] Microphone stream already active
[2026-07-20][06:10:12][handy_app_lib::actions][DEBUG] Failed to start recording: Recorder not available
[2026-07-20][06:10:12][handy_app_lib::transcription_coordinator][DEBUG] Start for 'transcribe' did not begin recording; staying idle
[2026-07-20][06:10:34][handy_app_lib::managers::audio][DEBUG] Microphone stream stopped
[2026-07-20][06:10:34][handy_app_lib::managers::audio][DEBUG] device resolve: no mic configured -> system default

2. Mid-Session Disconnect & Zero-Sample Channel Error:
[2026-07-20][20:43:32][handy_app_lib::audio_toolkit::audio::recorder][INFO] Using device: Ok("Microphone (USB Condenser Microphone)")
Sample rate: 48000
Channels: 2
Format: F32
[2026-07-20][20:43:37][handy_app_lib::audio_toolkit::audio::recorder][ERROR] Stream error: The requested device is no longer available. For example, it has been unplugged.
[2026-07-20][20:44:15][handy_app_lib::actions][DEBUG] TranscribeAction::stop called for binding: transcribe
[2026-07-20][20:44:15][handy_app_lib::managers::audio][ERROR] stop() failed: sending on a closed channel
[2026-07-20][20:44:15][handy_app_lib::actions][DEBUG] Recording stopped and samples retrieved in 241.4µs, sample count: 0
[2026-07-20][20:44:15][handy_app_lib::actions][DEBUG] Recording produced no audio samples; skipping persistence
[2026-07-20][20:44:33][handy_app_lib::actions][DEBUG] TranscribeAction::start called for binding: transcribe
[2026-07-20][20:44:33][handy_app_lib::managers::audio][DEBUG] Microphone stream already active
[2026-07-20][20:44:33][handy_app_lib::actions][DEBUG] Failed to start recording: Recorder not available
[2026-07-20][20:44:33][handy_app_lib::transcription_coordinator][DEBUG] Start for 'transcribe' did not begin recording; staying idle

```

--- a/src-tauri/src/audio_toolkit/audio/recorder.rs
+++ b/src-tauri/src/audio_toolkit/audio/recorder.rs
@@ -138,7 +138,14 @@ impl AudioRecorder {
 
     pub fn open(&mut self, device: Option<Device>) -> Result<(), Box<dyn std::error::Error>> {
         if self.worker_handle.is_some() {
-            return Ok(()); // already open
+            if !self.is_capture_worker_dead() {
+                return Ok(()); // already open
+            }
+            // The worker exited on its own (see `is_capture_worker_dead`). Reap
+            // it so we rebuild the stream below instead of handing the caller
+            // back a recorder whose channels are already closed.
+            log::warn!("Capture worker exited; rebuilding microphone stream");
+            let _ = self.close();
         }
 
         let (sample_tx, sample_rx) = mpsc::channel::<AudioChunk>();
@@ -331,6 +338,20 @@ impl AudioRecorder {
         Ok(resp_rx.recv()?) // wait for the samples
     }
 
+    /// True once the capture worker has exited without anyone calling `close`.
+    ///
+    /// `run_consumer` is driven entirely by the sample channel, so when cpal
+    /// tears the stream down mid-session (device unplugged, USB/Bluetooth
+    /// dropout) `sample_rx.recv()` returns `Err`, the loop ends and the worker
+    /// thread finishes. `cmd_tx` and `worker_handle` are still populated at
+    /// that point, so the recorder looks open from the outside while every
+    /// command sent to it fails on a closed channel.
+    pub fn is_capture_worker_dead(&self) -> bool {
+        self.worker_handle
+            .as_ref()
+            .is_some_and(|handle| handle.is_finished())
+    }
+
     pub fn close(&mut self) -> Result<(), Box<dyn std::error::Error>> {
         if let Some(tx) = self.cmd_tx.take() {
             let _ = tx.send(Cmd::Shutdown);
@@ -472,7 +493,16 @@ pub fn is_no_input_device_error(error_message: &str) -> bool {
 
 #[cfg(test)]
 mod tests {
-    use super::{is_microphone_access_denied, is_no_input_device_error};
+    use super::{is_microphone_access_denied, is_no_input_device_error, AudioRecorder};
+
+    #[test]
+    fn unopened_recorder_is_not_reported_dead() {
+        // No worker has been spawned yet, so there is nothing to reap. Guards
+        // against inverting the "no worker" case, which would make every first
+        // open() take the rebuild path.
+        let recorder = AudioRecorder::new().expect("recorder");
+        assert!(!recorder.is_capture_worker_dead());
+    }
 
     #[test]
     fn detects_access_is_denied() {

--- a/src-tauri/src/managers/audio.rs
+++ b/src-tauri/src/managers/audio.rs
@@ -10,7 +10,7 @@ use crate::helpers::clamshell;
 use crate::managers::transcription::StreamRouter;
 use crate::settings::{get_settings, AppSettings};
 use crate::utils;
-use log::{debug, error, info, warn};
+use log::{debug, error, info, trace, warn};
 use std::path::Path;
 use std::sync::atomic::{AtomicU64, Ordering};
 use std::sync::{Arc, Mutex};
@@ -532,7 +532,10 @@ impl AudioRecordingManager {
                 .is_some_and(|rec| rec.is_capture_worker_dead());
 
             if !worker_dead {
-                debug!("Microphone stream already active");
+                // trace, not debug: with the aliveness check in
+                // try_start_recording this now fires on every keypress in
+                // always-on mode.
+                trace!("Microphone stream already active");
                 return Ok(());
             }
 
@@ -680,15 +683,18 @@ impl AudioRecordingManager {
         let mut state = self.state.lock().unwrap();
 
         if let RecordingState::Idle = *state {
-            // Ensure microphone is open in on-demand mode
-            if matches!(*self.mode.lock().unwrap(), MicrophoneMode::OnDemand) {
-                // Cancel any pending lazy close
-                self.close_generation.fetch_add(1, Ordering::SeqCst);
-                if let Err(e) = self.start_microphone_stream() {
-                    let msg = format!("{e}");
-                    error!("Failed to open microphone stream: {msg}");
-                    return Err(msg);
-                }
+            // Cancel any pending lazy close (no-op in always-on mode, where
+            // closes are never scheduled).
+            self.close_generation.fetch_add(1, Ordering::SeqCst);
+            // Opens the stream in on-demand mode. In always-on mode the stream
+            // is normally already open and this is a cheap aliveness check —
+            // but if the capture worker died (device disconnect), it rebuilds
+            // the stream instead of leaving every subsequent start wedged on
+            // "Recorder not available".
+            if let Err(e) = self.start_microphone_stream() {
+                let msg = format!("{e}");
+                error!("Failed to open microphone stream: {msg}");
+                return Err(msg);
             }
 
             if let Some(rec) = self.recorder.lock().unwrap().as_ref() {

--- a/src-tauri/src/managers/audio.rs
+++ b/src-tauri/src/managers/audio.rs
@@ -518,8 +518,43 @@ impl AudioRecordingManager {
     pub fn start_microphone_stream(&self) -> Result<(), anyhow::Error> {
         let mut open_flag = self.is_open.lock().unwrap();
         if *open_flag {
-            debug!("Microphone stream already active");
-            return Ok(());
+            // `is_open` only records that we opened a stream at some point, not
+            // that one is still running. If the capture worker has since exited
+            // (mic unplugged mid-session, USB dropout), returning Ok here hands
+            // the caller a dead recorder: it captures nothing, then fails in
+            // stop() on the closed channel, and stays wedged until the
+            // on-demand close timeout eventually resets the manager.
+            let worker_dead = self
+                .recorder
+                .lock()
+                .unwrap()
+                .as_ref()
+                .is_some_and(|rec| rec.is_capture_worker_dead());
+
+            if !worker_dead {
+                debug!("Microphone stream already active");
+                return Ok(());
+            }
+
+            warn!("Microphone stream is no longer running (device disconnected?); reopening");
+
+            // Torn down inline rather than via stop_microphone_stream(), which
+            // takes the `is_open` lock we are already holding.
+            {
+                let mut mute_guard = self.mute_state.lock().unwrap();
+                if mute_guard.did_mute {
+                    restore_mute(mute_guard.prev_muted);
+                    mute_guard.did_mute = false;
+                }
+            }
+            if let Some(rec) = self.recorder.lock().unwrap().as_mut() {
+                // Skipping rec.stop() here: the worker is gone, so the command
+                // would only fail on the closed channel.
+                let _ = rec.close();
+            }
+            *self.is_recording.lock().unwrap() = false;
+            *open_flag = false;
+            // Fall through and open a fresh stream.
         }
 
         let start_time = Instant::now();

